### PR TITLE
Refactor Grist node

### PR DESCRIPTION
**Pull request by ivov**
Refactoring of #2157

### CODE REVIEW

**Comment by alexmojaki** [comment]
Wow, this is a big change! Going through it now.

File: packages/nodes-base/credentials/GristApi.credentials.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+export class GristApi implements ICredentialType {
+	name = 'gristApi';
+	displayName = 'Grist API';
+	documentationUrl = 'grist';
+	properties: INodeProperties[] = [
+		{
+			displayName: 'API Key',
+			name: 'apiKey',
+			type: 'string',
+			default: '',
```
Comment (alexmojaki) [inline packages/nodes-base/credentials/GristApi.credentials.ts:15]: This needs to be required (I see I made the same mistake)

File: packages/nodes-base/credentials/GristApi.credentials.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+					value: 'paid',
+				},
+			],
+		},
+		{
+			displayName: 'Custom Subdomain',
+			name: 'customSubdomain',
+			type: 'string',
+			default: '',
+			required: true,
```
Comment (alexmojaki) [inline packages/nodes-base/credentials/GristApi.credentials.ts:38]: Not sure if this is a bug in n8n, this says it's required but it's not actually forcing me to fill it in, and of course leaving it blank leads to other errors.

File: packages/nodes-base/nodes/Grist/Grist.node.ts [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+					const dataToSend = this.getNodeParameter('dataToSend', 0) as 'defineBelow' | 'autoMapInputData';
+
+					if (dataToSend === 'autoMapInputData') {
+						const incomingKeys = Object.keys(items[i].json);
+						const rawInputsToIgnore = this.getNodeParameter('inputsToIgnore', i) as string;
+						const inputDataToIgnore = rawInputsToIgnore.split(',').map(c => c.trim());
+
+						for (const key of incomingKeys) {
+							if (inputDataToIgnore.includes(key)) continue;
+							body.records[i] = { fields: { [key]: items[i].json[key] } };
```
Comment (alexmojaki) [inline packages/nodes-base/nodes/Grist/Grist.node.ts:100]: This doesn't work with multiple input items. The loop sends one record at a time, so when `i > 0` this sets all previous values in the array to null or undefined. I get an error showing this in the details:

```json
  "body": {
    "records": [
      null,
      {
        "fields": {
          "foo": 2
        }
      }
    ]
  }
```

File: packages/nodes-base/nodes/Grist/Grist.node.ts [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+							throw new NodeOperationError(this.getNode(), 'Please enter data to send');
+						}
+
+						body.records = [
+							{ fields: parseFieldsToSend(properties) },
+						];
+					}
+
+					const docId = this.getNodeParameter('docId', i) as string;
+					const tableId = this.getNodeParameter('tableId', i) as string;
```
Comment (alexmojaki) [inline packages/nodes-base/nodes/Grist/Grist.node.ts:118]: Should these be allowed to vary for different input items? `getTableColumns` doesn't which makes fields have fixed options. Are users allowed to enter an expression in `fieldId` which doesn't match those options?

File: packages/nodes-base/nodes/Grist/Grist.node.ts [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+						qs.filter = JSON.stringify(parsed);
+					}
+
+					responseData = await gristApiRequest.call(this, 'GET', endpoint, {}, qs);
+					responseData = responseData.records;
+				}
+
+			} catch (error) {
+				if (this.continueOnFail()) {
+					returnData.push({ json: { error: error.message } });
```
Comment (alexmojaki) [inline packages/nodes-base/nodes/Grist/Grist.node.ts:234]: It's not clear to me when `json` should be used if already using `returnJsonArray`? I tried a request which failed on one item and this is what the JSON output looks like:

```json
[
  {
    "id": 2
  },
  {
    "json": {
      "error": "The service was not able to process your request"
    }
  }
]
```

**Comment by alexmojaki** [comment]
Thanks. I'm going to start writing documentation now based on this PR, let me know if you think any part of this UI might change.

**Comment by ivov** [comment]
Confirmed, this will likely be the final UI. If anything is added, the docs team will make any necessary edits.

Thanks again for your hard work.

**Comment by dsagal** [comment]
I’d love to see this merged to use Grist with n8n!

File: packages/nodes-base/nodes/Grist/GenericFunctions.ts [hunk 6]
```diff
@@ -72,10 +71,10 @@ export function parseSortProperties(sortProperties: GristSortProperties) {
 }
 
 export function parseFilterProperties(filterProperties: GristFilterProperties) {
- return filterProperties.reduce<{ [key: string]: string[]; }>((acc, cur) => {
+	return filterProperties.reduce<{ [key: string]: string[]; }>((acc, cur) => {
 		acc[cur.field] = acc[cur.field] ?? [];
-		const values = cur.values.split(',').map(v => v.trim());
```
Comment (alexmojaki) [inline packages/nodes-base/nodes/Grist/GenericFunctions.ts]: If the values cannot be separated by commas, then there needs to be an update to:

- The docs https://github.com/n8n-io/n8n-docs/pull/619/files#diff-57ee899626d8d1192a3ec966c6a285f41905ac1007ef4684082b400d2cd7c542R115
- The field name, displayName, and description https://github.com/n8n-io/n8n/pull/2158/files#diff-1e5990e55548d4729b196e41b6b3894516c358970965696360ff203c16977a74R161

File: packages/nodes-base/nodes/Grist/GenericFunctions.ts [hunk 7]
```diff
@@ -72,10 +71,10 @@ export function parseSortProperties(sortProperties: GristSortProperties) {
...
-		acc[cur.field].push(...values);
+		const values = cur.values;
```
Comment (alexmojaki) [inline packages/nodes-base/nodes/Grist/GenericFunctions.ts:76]: If the values cannot be separated by commas, then there needs to be an update to:

- The docs https://github.com/n8n-io/n8n-docs/pull/619/files#diff-57ee899626d8d1192a3ec966c6a285f41905ac1007ef4684082b400d2cd7c542R115
- The field name, displayName, and description https://github.com/n8n-io/n8n/pull/2158/files#diff-1e5990e55548d4729b196e41b6b3894516c358970965696360ff203c16977a74R161

File: packages/nodes-base/nodes/Grist/Grist.node.ts [hunk 8]
```diff
@@ -68,6 +76,47 @@ export class Grist implements INodeType {
...
+
+				const options: OptionsWithUri = {
+					headers: {
+						Authorization: `Bearer ${apiKey}`,
+					},
+					method: 'GET',
+					uri: `https://${subdomain}.getgrist.com/api${endpoint}`,
+					qs: { limit: 1 },
+					body: {},
+					json: true,
```
Comment (alexmojaki) [inline packages/nodes-base/nodes/Grist/Grist.node.ts:103]: Since it's a GET, I think these can be removed.

**Comment by janober** [comment]
Got released with n8n@0.141.0

**Pull request merged**